text/0000-argument-lifetimes.md

```diff
@@ -0,0 +1,492 @@
+usually leave it off. The reason is that a lot of code works like the getter
+example above, where you're returning borrowed data which could only have come
+from the borrowed data you took in. Thus, in `get_field1` the lifetime for
+`&self` and for `&Foo` are assumed to be the same.
+
+Rust is conservative about leaving lifetimes off, though: if there's any
+ambiguity, you need to say explicitly state the relationships between the
+loans. So for example, the following function signature is *not* accepted:
+
+```rust
+fn select(data: &Data, params: &Params) -> &Item;
+```
+
+Rust cannot tell how long the resulting borrow of `Item` is valid for; it can't
+deduce its lifetime. Instead, you need to connect it to one or both of the input
+borrows:
+
+```rust
+fn select(data: &'data Data, params: &Params) -> &'data Item;
+fn select(data: &'both Data, params: &'both Params) -> &'both Item;
+```
+
+This notation lets you *name* the lifetime associated with a borrow and refer to
+it later:
+
+- In the first variant, we name the `Data` borrow lifetime `'data`, and make
+clear that the returned `Item` borrow is valid for the same lifetime.
+
+- In the second variant, we give *both* input lifetimes the *same* name `'both`,
+which is a way of asking the compiler to determine their "intersection"
+(i.e. the period for which both of the loans are active); we then say the
+returned `Item` borrow is valid for that period (which means it may incorporate
+data from both of the input borrows).
+
+You can use any names you like when introducing lifetimes (which always start
+with `'`), but are encouraged to make them meaningful, e.g. by using the same
+name as the parameter.
+
+## `struct`s and lifetimes
+
+Sometimes you need to build data types that contain borrowed data. Since those
+types can then be used in many contexts, you can't say in advance what the
+lifetime of those borrows will be. Instead, you must take it as a parameter:
+
+```rust
+struct VecIter<'vec, T> {
+    vec: &'vec Vec<T>,
+    index: usize,
+}
+```
+
+Here we're defining a type for iterating over a vector, without requiring
+*ownership* of that vector. To do so, we store a *borrow* of the vector. But
+because our new `VecIter` struct contains borrowed data, it needs to surface
+that fact, and the lifetime connected with it. It does so by taking an explicit
+`'vec` parameter for the relevant lifetime, and using it within.
+
+When using this struct, you can apply explicitly-named lifetimes as usual:
+
+```rust
+impl<T> Vec<T> {
+    fn iter(&'vec self) -> VecIter<'vec, T> { ... }
+}
+```
+
+However, in cases like this example, we would normally be able to leave off the
+lifetime with `&`, since there's only one source of data we could be borrowing
+from. We can do something similar with structs:
+
+```rust
+impl<T> Vec<T> {
+    fn iter(&self) -> VecIter<_, T> { ... }
+}
+```
+
+The `_` marker makes clear to the reader that *borrowing is happening*, which
+might not otherwise be clear.
```

Comment:
(On the other hand, `'_` *would* convey to me that a borrow is happening. I have a vague recollection that the token `'_` is painful to the eyes of some, though...)